examples/hetero/recommender_system.py

```diff
@@ -0,0 +1,400 @@
+
+
+model = Model(num_users=data['user'].num_nodes,
+              num_movies=data['movie'].num_nodes,
+              hidden_channels=64).to(device)
+optimizer = torch.optim.AdamW(model.parameters(), lr=2e-5)
+
+
+def get_user_positive_items(edge_index):
+    # Get all the movies that the user has actually watched
+    user_pos_items = {}
+    for i in range(edge_index.shape[1]):
+        user = edge_index[0][i].item()
+        item = edge_index[1][i].item()
+        if user not in user_pos_items:
+            user_pos_items[user] = []
+        user_pos_items[user].append(item)
+    return user_pos_items
+
+
+def write_recs(real_recs, val_data, val_user_pos_items):
+    movie_path = osp.join(path, './raw/ml-latest-small/movies.csv')
+
+    # load user and movie nodes
+
+    def load_node_csv(path, index_col):
+        df = pd.read_csv(path, index_col=index_col)
+        return df
+
+    movie_id_to_name = load_node_csv(movie_path, index_col='movieId')
+    with open('recommendations.txt', 'w') as f:
+        for user in range(val_data['user'].num_nodes):
+            try:
+                if user in val_user_pos_items:
+                    f.write('\nPred\n')
+                    f.write(
+                        json.dumps(
+                            movie_id_to_name.iloc[real_recs[user]].to_dict()))
+                    f.write('\nGT\n')
+                    f.write(
+                        json.dumps(movie_id_to_name.iloc[
+                            val_user_pos_items[user]].to_dict()))
+            except KeyError:
+                pass
+
+
+def get_embeddings(model, val_data):
+    # Get user and movie embeddings from the model
+    x_dict = {}
+    x_dict['user'] = model.user_emb(val_data['user'].n_id)
+    x_dict['movie'] = torch.cat(
+        [model.movie_emb(val_data['movie'].n_id), val_data.x_dict['movie']],
+        dim=-1)
+    embs = model.encoder(x_dict, val_data.edge_index_dict)
+    return embs['movie'], embs['user']
+
+
+def make_recommendations(model, train_data, val_data, k, write_to_file=False):
+
+    movie_embs, user_embs = get_embeddings(model, train_data)
+    mipsknn = MIPSKNNIndex(movie_embs)
+    knn_score, knn_index = mipsknn.search(
+        user_embs, k, train_data['user', 'movie'].edge_index)
+
+    # Obtain the movies that each user has already
+    # watched. These movies need to be removed from
+    # recommendations
+    user_pos_items = get_user_positive_items(train_data['user',
+                                                        'movie'].edge_index)
+
+    # Get the recommendations for user; only those
+    # that the user is yet to watch
+    real_recs = {}
+    for user in range(train_data['user'].num_nodes):
+        try:
+            real_recs[user] = list(
+                OrderedSet(knn_index[user].tolist()) -
+                OrderedSet(user_pos_items[user]))[:k]
+        except KeyError:
+            real_recs[user] = knn_index[user].tolist()[:k]
```

Comment:
I think this example looks mostly good, thank you :) The only remaining issue I have is in this code-block. I think I would prefer if we can do `mipsknn.search` in mini-batches over users, and then also compute the metrics in a mini-batch fashion. Let me know if you need help on that.

Reply:
Implemented in the latest commit. Thanks